tutorials/haskell/dotenv-hs/tutorial.md

```diff
@@ -0,0 +1,284 @@
+  -x,--dotenv-example DOTENV_EXAMPLE
+                           File with all the necesary env
+                           variables (default: ".env.example")
+  -o,--override            Override existing variables
+```
+
+The minimum necesary is a **PROGRAM** to run. The paths for the *dotenv file*
+and the *dotenv example file* can be set with the short flags `-e` and `-x`,
+respectivelly. The default values are *.env* and *.env.example*, respectivelly.
+The env vars defined in the environment can be overrided with the ones defined
+in the *dotenv file* using the `-o` flag.
+
+Now, go to the directory of the example and follow the next instructions:
+
+Install dependencies and build the project:
+
+```shell
+$ stack build
+```
+
+or
+
+```shell
+$ cabal install --only-dependencies
+$ cabal build
+```
+
+Two executable applications will be created: `dotenv-st` and `dotenv-code`. The
+first one will be used for this part of the example.
+
+Create the databases, one for development and the other one for testing:
+
+```shell
+$ createdb dotenv-dev
+$ createdb dotenv-test
+```
+
+Load the data:
+
+```shell
+$ psql -d dotenv-dev  < sql/dotenv-dev.sql
+$ psql -d dotenv-test < sql/dotenv-test.sql
+```
+
+Before executing the application, check out the *.env.example* file.
+
+```shell
+$ cat .env.example
+
+DB_HOST=
+DB_PORT=
+DB_NAME=
+DB_USER=
+DB_PASS=
+```
+
+This file contains all the *necesary env vars* for the configuration of the
+connection with the database: the host, the port, the name of the database,
+the username and its password. Without these variables defined the application
+will be stopped before start.
+
+Also, check the *.env* file. The *.env.test* file is the same, except by the
+`DB_NAME`.
+
+```shell
+$ cat .env
+
+DB_HOST=localhost
+DB_PORT=5432
+DB_NAME=dotenv-dev
+DB_USER=postgres
+DB_PASS=postgres
+```
+
+The application will read this environment variables and get the connection
+information to connect with the database. As the below code shows:
+
+```haskell
+readConfig :: IO ConnectInfo
+readConfig =
+  ConnectInfo
+    <$> getEnv "DB_HOST"
+    <*> fmap read (getEnv "DB_PORT")
+    <*> getEnv "DB_NAME"
+    <*> getEnv "DB_USER"
+    <*> getEnv "DB_PASS"
+```
+
+Run the `dotenv-st` application:
+
+```shell
+$ stack exec dotenv-st
+dotenv-st: DB_HOST: getEnv: does not exist (no environment variable)
+```
+
+The error displayed in the screen says that the `DB_HOST` environment variable
+must exist before executing the application. Let's execute it again, but
+this time using `dotenv`.
+
+```shell
+$ dotenv "stack exec dotenv-st"
+User {id_ = 1, name = "User 1", lastname = "User 1", birthDate = 1990-01-01}
+User {id_ = 2, name = "User 2", lastname = "User 2", birthDate = 1990-01-01}
+User {id_ = 3, name = "User 3", lastname = "User 3", birthDate = 1990-01-01}
+```
+
+The application was executed without any problem. Awesome! Let's run the
+tests for the application with `dotenv`.
+
+```shell
+$ dotenv "stack test"
```

Comment:
One of the things that I expressed in this tutorial is the importance about the separating environments (development, testing, and so on). I wrote the tests with a different *.env.example* to show this separation. So, I'll stick with the tests. However, we can discuss it.